File: CombatCore/PlayerState.cs
namespace CombatCore;

/// <summary>
/// Everything about one player that participates in the simulation. Deliberately
/// a struct: MatchState (which holds two of these) needs to be cheaply copyable
/// so the rollback controller can keep a ring buffer of past snapshots without
/// allocating per-frame.
/// </summary>
public struct PlayerState
{
    public FixedVector3 Position;
    public int FacingSign;          // +1 = facing +X, -1 = facing -X
    public int Health;
    public int MaxHealth;

    public MoveId CurrentMove;
    public int MoveFrame;           // frames elapsed since CurrentMove started (0-based)

    public int HitstunFramesRemaining;
    public int BlockstunFramesRemaining;

    public bool IsHoldingBlock;
    public bool IsBeastForm;

    // Prevents a single multi-frame active hitbox window (e.g. HeavyPunch's
    // 4-frame active window) from registering as 4 separate hits against the
    // same activation. Reset to false whenever a new attack move starts.
    public bool HasCurrentAttackConnected;

    public static PlayerState CreateDefault(FixedVector3 startPosition, int facingSign, int maxHealth = 100)
    {
        return new PlayerState
        {
            Position = startPosition,
            FacingSign = facingSign,
            Health = maxHealth,
            MaxHealth = maxHealth,
            CurrentMove = MoveId.Idle,
            MoveFrame = 0,
            HitstunFramesRemaining = 0,
            BlockstunFramesRemaining = 0,
            IsHoldingBlock = false,
            IsBeastForm = false,
        };
    }

    public readonly bool IsKnockedOut => Health <= 0;
    public readonly bool IsInHitstun => HitstunFramesRemaining > 0;
    public readonly bool IsInBlockstun => BlockstunFramesRemaining > 0;
    public readonly bool CanAct => !IsInHitstun && !IsInBlockstun && !IsKnockedOut;

    public readonly MoveDefinition MoveDef => MoveTable.Moves[CurrentMove];
}

File: Client/src/sim/PlayerState.ts
import { FixedVector3 } from "./FixedVector3.ts";
import { MoveId } from "./FrameData.ts";
import type { MoveDefinition, MoveIdValue } from "./FrameData.ts";
import { getMoveDef } from "./MoveTable.ts";

export interface PlayerState {
  position: FixedVector3;
  facingSign: number;
  health: number;
  maxHealth: number;

  currentMove: MoveIdValue;
  moveFrame: number;

  hitstunFramesRemaining: number;
  blockstunFramesRemaining: number;

  isHoldingBlock: boolean;
  isBeastForm: boolean;

  // Prevents a single multi-frame active hitbox window (e.g. HeavyPunch's
  // 4-frame active window) from registering as 4 separate hits against the
  // same activation. Reset to false whenever a new attack move starts.
  hasCurrentAttackConnected: boolean;
}

export function createDefaultPlayerState(startPosition: FixedVector3, facingSign: number, maxHealth = 100): PlayerState {
  return {
    position: startPosition,
    facingSign,
    health: maxHealth,
    maxHealth,
    currentMove: MoveId.Idle,
    moveFrame: 0,
    hitstunFramesRemaining: 0,
    blockstunFramesRemaining: 0,
    isHoldingBlock: false,
    isBeastForm: false,
    hasCurrentAttackConnected: false,
  };
}

export function isKnockedOut(p: PlayerState): boolean {
  return p.health <= 0;
}

export function moveDef(p: PlayerState): MoveDefinition {
  return getMoveDef(p.currentMove);
}

/** Structural clone - PlayerState is a plain object standing in for C#'s value-type struct semantics. */
export function clonePlayerState(p: PlayerState): PlayerState {
  return { ...p, position: p.position };
}
